File: leetcodePython/Reverse_Integer.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""
Created on Wed Mar  7 12:05:34 2018

@author: junchiliu

Given a 32-bit signed integer, 
reverse digits of an integer.
Note:
Assume we are dealing with an environment 
which could only hold integers 
within the 32-bit signed integer range. 
For the purpose of this problem, 
assume that your function 
returns 0 when the reversed integer overflows.
"""

def reverse(x):
        if x>=0:
            a = int(str(x)[::-1])
            if a >=2**31:
                return 0
            else:
                return a
        else:
            x=abs(x)
            a = int(str(x)[::-1])
            if a >2**31:
                return 0 
            else:
                return a*(-1)
    

File: leetcodeGolang/reverseInteger.go
/*
Given a 32-bit signed integer, reverse digits of an integer.
Note:
Assume we are dealing with an environment which could only store integers within 
the 32-bit signed integer range: [−231,  231 − 1]. 
For the purpose of this problem, 
assume that your function returns 0 when the reversed integer overflows.

Example 1:

Input: 123
Output: 321
Example 2:

Input: -123
Output: -321
Example 3:

Input: 120
Output: 21
*/

package main

import (
	"math"
	"fmt"
)

func reverse(x int) int {
	sign := 1

	if x < 0 {
		sign = -1
		x = -1 * x
	}

	res := 0
	for x > 0 {
		temp := x % 10
		res = res * 10 + temp
		x = x / 10
	}
	
	res = sign * res

	if res > math.MaxInt32 || res < math.MinInt32 {
		res = 0
	}

	return res
}

func main()  {

	a := 321
	b := reverse(a)
	fmt.Println(b)
	
}